docarray/array/array.py

```diff
@@ -13,34 +38,326 @@ class DocumentArray(
     BaseNode,
 ):
     """
-    a DocumentArray is a list-like container of Document of the same schema
+    A DocumentArray is a container of Documents.
 
     :param docs: iterable of Document
+
+    A DocumentArray is a list of Documents of any schema. However, many
+    DocumentArray features are only available if these Documents are
+    homogeneous and follow the same schema. To be precise, in this schema you can use
+    the `DocumentArray[MyDocument]` syntax where MyDocument is a Document class
+    (i.e. schema). This creates a DocumentArray that can only contain Documents of
+    the type 'MyDocument'.
+
+    EXAMPLE USAGE
+    .. code-block:: python
+        from docarray import Document, DocumentArray
+        from docarray.typing import NdArray, ImageUrl
+
+
+        class Image(Document):
+            tensor: Optional[NdArray[100]]
+            url: ImageUrl
+
+
+        da = DocumentArray[Image](
+            Image(url='http://url.com/foo.png') for _ in range(10)
+        )  # noqa: E510
+
+
+    If your DocumentArray is homogeneous (i.e. follows the same schema), you can access
+    the field at the DocumentArray level (for example `da.tensor`). You can also set
+    them, with `da.tensor = np.random.random([10, 100])`
+
+
+    A DocumentArray can be in one of two modes: unstacked mode and stacked mode.
+
+
+    **Unstacked mode (default)**:
+    In this case a DocumentArray is a list of Documents and each Document owns its data.
+    The getter and setter shown above returns a list of the fields of each Document
+    (or a DocumentArray if the field is a nested Document). This list/DocumentArray
+    is created on the fly. The setter sets the field of each Document to the value
+    of the list/DocumentArray/Tensor passed as parameters.
+
+    This list-like behavior is not always optimal, especially when you want
+    to process data in batches or perform operations involving matrix computation.
+    This is where the stack mode of the DocumentArray comes in handy.
+
+    In **stacked mode** tensor-like fields of every Document are stored on the
+    DocumentArray level, as one stacked tensor. This enables operations on the entire
+    batch without iterating over the DocumentArray.
+    In this mode the Documents in in the Document don't own the data anymore but just
+    reference the data in the DocumentArray's tensor.
+    Operations like `da.append` are not supported because
+    they are too slow. For these operations you should use unstacked
+    mode instead.
```

Comment:
```suggestion
    In **stacked mode**, tensor-like fields of every Document are stored on the
    DocumentArray level as one stacked tensor. This enables operations on the entire
    batch without iterating over the DocumentArray.
    In this mode the Documents in the DocumentArray no longer own the data but just
    reference the data in the DocumentArray's tensor.
    Operations like `da.append` are not supported because
    they are too slow. For these operations you should use unstacked
    mode instead.
```